test/helpers/utils.go

```diff
@@ -1157,6 +1162,146 @@ func DeleteNodeAndMachine(t *testing.T, cs *framework.ClientSet, node corev1.Nod
 	require.NoError(t, delErr)
 }
 
+// Sets the MachineSet to the desired number of replicas. Note: Does not wait
+// for new nodes to become available nor does not wait for nodes to be scaled
+// back down. Returns a function that will scale the nodes back to the original
+// MachineSet replica value.
+func ScaleMachineSet(t *testing.T, cs *framework.ClientSet, machinesetName string, desiredReplicas int32) func() {
+	t.Helper()
+
+	machineclient := machineClientv1beta1.NewForConfigOrDie(cs.GetRestConfig())
+
+	machineset, err := machineclient.MachineSets(machineAPINamespace).Get(context.TODO(), machinesetName, metav1.GetOptions{})
+	require.NoError(t, err)
+
+	originalReplicaCount := *machineset.Spec.Replicas
+	t.Logf("MachineSet %s has original replica count of %d. Scaling to %d replicas.", machinesetName, originalReplicaCount, desiredReplicas)
+
+	machineset.Spec.Replicas = &desiredReplicas
+
+	_, err = machineclient.MachineSets(machineAPINamespace).Update(context.TODO(), machineset, metav1.UpdateOptions{})
+	require.NoError(t, err)
+
+	return func() {
+		machineset, err := machineclient.MachineSets(machineAPINamespace).Get(context.TODO(), machinesetName, metav1.GetOptions{})
+		require.NoError(t, err)
+
+		machineset.Spec.Replicas = &originalReplicaCount
+
+		_, err = machineclient.MachineSets(machineAPINamespace).Update(context.TODO(), machineset, metav1.UpdateOptions{})
+		require.NoError(t, err)
+
+		t.Logf("Scaled MachineSet %s back to the original replica count of %d", machinesetName, originalReplicaCount)
+	}
+}
+
+// Performs the same task as ScaleMachineSet, but waits for the new nodes to
+// all be ready before returning. Returns a list of new nodes as well as a
+// function that will scale the nodes back down and delete the underlying
+// machines / nodes.
+func ScaleMachineSetAndWaitForNodesToBeReady(t *testing.T, cs *framework.ClientSet, machinesetName string, desiredReplicas int32) ([]*corev1.Node, func()) {
+	nodes, err := cs.CoreV1Interface.Nodes().List(context.TODO(), metav1.ListOptions{})
+	require.NoError(t, err)
+
+	// Get a set of current node names before we scale up any additional nodes.
+	originalNodes := sets.New[string]()
+	for _, node := range nodes.Items {
+		originalNodes.Insert(node.Name)
+	}
+
+	start := time.Now()
+
+	// Start the scale-up process
+	scaledownFunc := ScaleMachineSet(t, cs, machinesetName, desiredReplicas)
```

Comment:
Isn't this cleanup function just scaling to the scaled up value?

Reply:
No. Before scaling up the MachineSet, we get the current desiredReplicas value. When the cleanup function is called, we go back to the original desiredReplicas value.

Reply:
> For most of our deployments, each machineset only has a default of 1 replica.

You're right. My sandbox cluster configs do it the other way though because there was a capacity constraint one time and it was easier to do it that way.

Regardless of how many nodes are in the MachineSet though, the test no longer hard-codes the `desiredReplicas` value. Instead, the only value that is hard-coded is how many additional nodes to scale :smile: .

Reply:
Oh, I think I understand now, we're tracking for originalReplicaCount, so the desired value doesn't matter in this case.